File: packages/backend/sdk/src/webhooks.ts
import { createHmac, timingSafeEqual } from 'node:crypto';

const MAX_AGE_SECONDS = 10;

function sortKeys(value: unknown): unknown {
  if (Array.isArray(value)) {
    return value.map(sortKeys);
  }
  if (value !== null && typeof value === 'object') {
    return Object.keys(value as object)
      .sort()
      .reduce(
        (acc, key) => {
          acc[key] = sortKeys((value as Record<string, unknown>)[key]);
          return acc;
        },
        {} as Record<string, unknown>,
      );
  }
  return value;
}

/**
 * Verify a Fluxis webhook signature.
 *
 * Fluxis signs webhook payloads using HMAC-SHA256 over
 * `"<timestamp>.<canonical_json>"`, where canonical JSON is the payload
 * with all object keys sorted recursively.
 *
 * Pass the parsed payload, the signature from `x-fluxis-signature`,
 * the timestamp from `x-fluxis-timestamp`, and the webhook secret obtained
 * when creating notification settings.
 */
export function verifyWebhookSignature(
  payload: unknown,
  signature: string,
  timestamp: string,
  secret: string,
): boolean {
  const requestTimestamp = parseInt(timestamp, 10);
  const now = Math.floor(Date.now() / 1000);

  if (Number.isNaN(requestTimestamp) || now - requestTimestamp > MAX_AGE_SECONDS) {
    return false;
  }

  const canonicalJson = JSON.stringify(sortKeys(payload));
  const signedString = `${timestamp}.${canonicalJson}`;
  const expectedSignature = createHmac('sha256', secret)
    .update(signedString)
    .digest('hex');

  if (signature.length !== expectedSignature.length) {
    return false;
  }

  return timingSafeEqual(
    Buffer.from(signature),
    Buffer.from(expectedSignature),
  );
}

File: packages/backend/sdk-python/src/fluxis/webhooks.py
"""Webhook signature verification utilities."""

from __future__ import annotations

import hashlib
import hmac
import json
import time
from collections.abc import Mapping

MAX_AGE_SECONDS = 10


def _sort_keys(value: object) -> object:
    if isinstance(value, list):
        return [_sort_keys(item) for item in value]
    if isinstance(value, Mapping):
        return {key: _sort_keys(value[key]) for key in sorted(value)}
    return value


def verify_webhook_signature(
    payload: object,
    signature: str,
    timestamp: str,
    secret: str,
) -> bool:
    """Verify a Fluxis webhook HMAC-SHA256 signature."""
    try:
        request_timestamp = int(timestamp)
    except ValueError:
        return False

    now = int(time.time())
    if now - request_timestamp > MAX_AGE_SECONDS:
        return False

    canonical_json = json.dumps(_sort_keys(payload), separators=(",", ":"))
    signed_string = f"{timestamp}.{canonical_json}"
    expected_signature = hmac.new(
        secret.encode("utf-8"),
        signed_string.encode("utf-8"),
        hashlib.sha256,
    ).hexdigest()

    if len(signature) != len(expected_signature):
        return False

    return hmac.compare_digest(signature, expected_signature)
